File: Java/24_DP/LC_790_DominoAndTrominoTiling.java
class Solution {

    private static final long MOD = (long) 1e9 + 7;

    public int numTilings(int n) {
        if (n == 1){
            return 1;
        }
            
        long[] dp = new long[n + 1];
        dp[0] = 1;
        dp[1] = 1;
        dp[2] = 2;
        for (int i = 3; i <= n; i++) {
            dp[i] = (dp[i - 1] * 2 + dp[i - 3]) % MOD;
        }

        return (int) dp[n];
    }
}

public class LC_790_DominoAndTrominoTiling {

}

File: Swift/Leetcode.playground/Sources/24_DP/LC_790_DominoAndTrominoTiling.swift
class Solution {
    private let MOD: Int64 = 1_000_000_007
    
    func numTilings(_ n: Int) -> Int {
        
        var dp = [Int](repeating: 0, count: 1001)
        dp[0] = 1
        dp[1] = 1
        dp[2] = 2
        
        if n > 2{
            for i in 3...n {
                dp[i] = (dp[i - 1] * 2 + dp[i - 3]) % 1000000007
            }
        }
        
        return dp[n]
    }
}